clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -848,6 +916,225 @@ void BuiltinNameEmitter::EmitQualTypeFinder() {
+    switch (ElementType) {
+    case BOOL:
+      TypeStr << "_bool_t";
+      break;
+    case INT:
+      TypeStr << "_int" << ElementSize << "_t";
+      break;
+    case UINT:
+      TypeStr << "_uint" << ElementSize << "_t";
+      break;
+    case FLOAT:
+      TypeStr << "_fp" << ElementSize << "_t";
+      break;
+    case BUILTIN:
+      TypeStr << "_" << Name;
+      break;
+    }
+  }
+
+  return static_cast<std::string>(Buffer);
+}
+
+void JSONBuiltinInterfaceEmitter::Emit() {
+
+  // Parse the Records to populate the internal lists.
+  GetOverloads();
+  GroupBySignature();
+  ExpandTypes();
+
+  EmitBuiltins();
+}
+
+// Generate all possible types represented by the input record.
+void JSONBuiltinInterfaceEmitter::ExpandType(const Record *Ty) {
+  auto ExpansionIt = ExpandedTypes.find(Ty);
+
+  if (ExpansionIt != ExpandedTypes.end())
+    return;
+  llvm::SmallVectorImpl<std::string> &Expansion = ExpandedTypes[Ty];
+
+  JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(Ty);
+
+  bool IsCompound = Ty->isSubClassOf("CompoundType");
+  if (IsCompound) {
+    const Record *EltRecord = Ty->getValueAsDef("ElementType");
+    ExpandType(EltRecord);
+    Expansion = ExpandedTypes[EltRecord];
+    for (std::string &TypeStr : Expansion) {
+      if (TypeDesc.IsPointer) {
+        if (TypeDesc.IsConst)
+          TypeStr += " const";
+        if (TypeDesc.IsVolatile)
+          TypeStr += " volatile";
+        TypeStr += " ";
+        TypeStr += TypeDesc.AddrSpace;
+        TypeStr += " *";
+      }
+    }
+
+    return;
+  }
+
+  if (Ty->isSubClassOf("GenericType")) {
+    assert(!Expansion.size() && "Types already expanded");
+    std::vector<long int> NVec =
+        Ty->getValueAsDef("VectorList")->getValueAsListOfInts("List");
+    std::vector<Record *> NTypes =
+        Ty->getValueAsDef("TypeList")->getValueAsListOfDefs("List");
+    for (const Record *SubT : NTypes) {
+      for (long int VecWidth : NVec) {
+        JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(SubT);
+        TypeDesc.VecWidth = VecWidth;
+        Expansion.emplace_back(TypeDesc.GetBaseTypeAsStr());
+      }
+    }
+
+    return;
+  }
+
+  Expansion.emplace_back(TypeDesc.GetBaseTypeAsStr());
+}
+
+void JSONBuiltinInterfaceEmitter::ExpandTypes() {
+  for (const Record *T : Records.getAllDerivedDefinitions("Type")) {
+    ExpandType(T);
+  }
+}
+
+void JSONBuiltinInterfaceEmitter::EmitBuiltins() {
+  struct FnDesc {
+    FnDesc(const Record *Rec)
+        : IsPure(Rec->getValueAsBit("IsPure")),
+          IsConst(Rec->getValueAsBit("IsConst")),
+          IsConv(Rec->getValueAsBit("IsConv")),
+          IsVariadic(Rec->getValueAsBit("IsVariadic")) {}
+
+    bool IsPure;
+    bool IsConst;
+    bool IsConv;
+    bool IsVariadic;
+    SmallVector<llvm::StringRef, 8> Args;
+  };
+  StringMap<SmallVector<FnDesc, 16>> NameToProtoList;
+
+  // For each function names, gather the list of overloads
+  for (const auto &SLM : SignatureListMap) {
```

Comment:
So I don't think auto fits in the range-for down this chain, the rest of your uses of auto seem to improve readability.

Reply:
made it verbose.